File: CS/WindowsApplication3/Selection.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraVerticalGrid;
using System.Diagnostics;
using System.Collections;
using DevExpress.XtraVerticalGrid.Rows;
using System.Linq;
using DevExpress.LookAndFeel;

namespace DXSample
{
    public class Selection : CollectionBase
    {
        private GridCell _StartRangeCell;
        VGridControlBase vgrid;
        public Selection(VGridControlBase vgrid)
        {
            this.vgrid = vgrid;
        }

        protected override void OnInsertComplete(int index, object value)
        {
            base.OnInsertComplete(index, value);
            InvalidateRow(value);
        }

        private void InvalidateRow(object value)
        {
            GridCell cell = value as GridCell;
            vgrid.InvalidateRow(cell.Row);
        }

        protected override void OnRemoveComplete(int index, object value)
        {
            base.OnRemoveComplete(index, value);
            InvalidateRow(value);
        }

        protected override void OnClear()
        {
            while (List.Count > 0)
                List.RemoveAt(List.Count - 1);
        }

        public bool Contains(GridCell cell)
        {
            if (cell.Row is CategoryRow)
                return List.OfType<GridCell>().Any(c => c.Row.Equals(cell.Row)); 
           return List.Contains(cell);
        }

        public void Add(GridCell cell)
        {
            List.Add(cell);
        }

        public void Remove(GridCell cell)
        {
            if (cell.Row is CategoryRow)
                List.Remove(List.OfType<GridCell>().First(c => c.Row.Equals(cell.Row)));
            else
                List.Remove(cell);
        }

        public GridCell this[int index]
        {
            get
            {
                return List[index] as GridCell;
            }
            set
            {
                List[index] = value;
            }
        }

        public int IndexOf(GridCell cell)
        {
            return List.IndexOf(cell);
        }


        public GridCell StartRangeCell
        {
            get { return _StartRangeCell; }
            set { _StartRangeCell = value; }
        }

    }
}

File: VB/WindowsApplication3/Selection.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraVerticalGrid
Imports System.Diagnostics
Imports System.Collections
Imports DevExpress.XtraVerticalGrid.Rows
Imports System.Linq
Imports DevExpress.LookAndFeel

Namespace DXSample
	Public Class Selection
		Inherits CollectionBase
		Private _StartRangeCell As GridCell
		Private vgrid As VGridControlBase
		Public Sub New(ByVal vgrid As VGridControlBase)
			Me.vgrid = vgrid
		End Sub

		Protected Overrides Sub OnInsertComplete(ByVal index As Integer, ByVal value As Object)
			MyBase.OnInsertComplete(index, value)
			InvalidateRow(value)
		End Sub

		Private Sub InvalidateRow(ByVal value As Object)
			Dim cell As GridCell = TryCast(value, GridCell)
			vgrid.InvalidateRow(cell.Row)
		End Sub

		Protected Overrides Sub OnRemoveComplete(ByVal index As Integer, ByVal value As Object)
			MyBase.OnRemoveComplete(index, value)
			InvalidateRow(value)
		End Sub

		Protected Overrides Sub OnClear()
			Do While List.Count > 0
				List.RemoveAt(List.Count - 1)
			Loop
		End Sub

		Public Function Contains(ByVal cell As GridCell) As Boolean
			If TypeOf cell.Row Is CategoryRow Then
				Return List.OfType(Of GridCell)().Any(Function(c) c.Row.Equals(cell.Row))
			End If
		   Return List.Contains(cell)
		End Function

		Public Sub Add(ByVal cell As GridCell)
			List.Add(cell)
		End Sub

		Public Sub Remove(ByVal cell As GridCell)
			If TypeOf cell.Row Is CategoryRow Then
				List.Remove(List.OfType(Of GridCell)().First(Function(c) c.Row.Equals(cell.Row)))
			Else
				List.Remove(cell)
			End If
		End Sub

		Default Public Property Item(ByVal index As Integer) As GridCell
			Get
				Return TryCast(List(index), GridCell)
			End Get
			Set(ByVal value As GridCell)
				List(index) = value
			End Set
		End Property

		Public Function IndexOf(ByVal cell As GridCell) As Integer
			Return List.IndexOf(cell)
		End Function


		Public Property StartRangeCell() As GridCell
			Get
				Return _StartRangeCell
			End Get
			Set(ByVal value As GridCell)
				_StartRangeCell = value
			End Set
		End Property

	End Class
End Namespace
